### DISCUSSION TITLE
Dockerfile setup for this project

### DISCUSSION THREAD

**Discussion by RobinsonLaw** [alejandroautalan/pygubu-designer#186] [Q&A] [answered]
Hi
I try the example success with VENV success. And how can it build the docker image with this project structure?
I only use the simple format of "Dockerfile" build the image to run. Program can be run but the tk lib not working.

FROM python:3.9

WORKDIR /app

COPY requirements.txt requirements.txt
RUN pip install -r requirements.txt

COPY . .

CMD ["python","app.py"]

https://github.com/alejandroautalan/pygubu/wiki/DevHowto/93dd4f04778234d145bbaa7de40e7a4ad9741229

**Comment by alejandroautalan** [comment]
Hello @RobinsonLaw, thanks for trying pygubu.

I have been using docker for test pygubu-designer in different python versions.

https://gist.github.com/alejandroautalan/db5d19fedffaf25aa41a5ce40af2b50f

Let me know if this helps you.

Regards
Alejandro A.

**Comment by RobinsonLaw** [comment]
I only use my program with import pygubu and coding. The examples of
pygubu-designer program are tools only for me. The Dockerfile setup may be
referenced but it is complicated.

在 2023年1月13日週五 05:03，Alejandro Autalán ***@***.***> 寫道：

**Comment by RobinsonLaw** [comment]
running inside the docker have the error message

Exception has occurred: TclError

   -

no display name and no $DISPLAY environment variable
File "/mypb/src/pygubu/src/pygubu/plugins/tk/tkstdwidgets.py", line 73, in
realize self.widget = tk.Tk() File
"/mypb/src/pygubu/src/pygubu/builder.py", line 210, in _realize
parent.realize(master, extra_init_args) File
"/mypb/src/pygubu/src/pygubu/builder.py", line 150, in get_object bobject =
self._realize(root, wmeta, extra_init_args) File
"/workspaces/mypygibu-pb/Premium/myUi/operation.py", line 37, in __init__
self.mainwindow = builder.get_object("toplevel1", master) File
"/workspaces/mypygibu-pb/Premium/app.py", line 83, in <module> app=
operation.Testui01App(app=setting,ui=ui_list)



Alejandro Autalán ***@***.***> 於 2023年1月13日週五 上午5:03寫道：

**Reply by alejandroautalan** [reply]
Hello, here are some tips I didn't mention before.

* Make sure you are running docker in your local computer. Running in a remote server will require different steps.
* tkinter needs a graphics server and a display. This example will use your current display server.
* To use your display server you need to setup the environment variables, and pass them to docker.
* Before you app starts, inside the container, receive the environment vars and setup the display using the xauth command.
* You need to mount /tmp/.X11-unix inside container, so tkinter can access your display.
* You need to run container with "network host" mode.
* You need xauth program installed in container and you computer.

As an example I will run pygubu-designer inside a test container, with manual steps.  Running on xubuntu 22.04.


https://user-images.githubusercontent.com/3482471/212237136-4369c41f-3f5f-44d9-877c-7ab6839054ce.mp4



If you need more help, please provide more info: full docker command, OS running, etc.

Regards
Alejandro A

**Comment by RobinsonLaw** [comment]
Hi
Sorry, I cannot understand your explanation in your example.
I am not using remote server and all run in local PC windows. Your case is
different from me. According to your example for virtual environment
https://github.com/alejandroautalan/pygubu/wiki/DevHowto/93dd4f04778234d145bbaa7de40e7a4ad9741229


1.  I made a project structure below

mypygibu-pb/
|--------pygubu
|          |-----src
|                   |-----pygubu
|                             |  builder.py
|--------venv
|--------Premium #my source code
|             |  app.py                        #  main start
|             |--------UI
|                         | operation.py    #call pygubu lib
|                         | test.ui              #
| Dockerfile
| requirements.txt


2. Dockerfile Content shows below:

FROM python:3.9

WORKDIR /mypb

COPY requirements.txt requirements.txt
RUN pip install -r requirements.txt

COPY ./Premium ./Premium

CMD ["python","./Premium/app.py"]


3. command to build in Dockerfile directory:  docker build -t  docker_name .

4. docker run  docker_name

5.  Error Show below:
Exception has occurred: TclError

   -

no display name and no $DISPLAY environment variable
File "/mypb/src/pygubu/src/pygubu/plugins/tk/tkstdwidgets.py", line 73, in
realize self.widget = tk.Tk() File
"/mypb/src/pygubu/src/pygubu/builder.py", line 210, in _realize
parent.realize(master, extra_init_args) File
"/mypb/src/pygubu/src/pygubu/builder.py", line 150, in get_object bobject =
self._realize(root, wmeta, extra_init_args) File
"/workspaces/mypygibu-pb/Premium/myUi/operation.py", line 37, in __init__
self.mainwindow = builder.get_object("toplevel1", master) File
"/workspaces/mypygibu-pb/Premium/app.py", line 83, in <module> app=
operation.Testui01App(app=setting,ui=ui_list)
.
6. Question:
docker Path is different from the original
"/mypb/ *src/pygubu/src/pygubu/**plugins/tk/tkstdwidgets.py*"

original path should be: mypygibu-pb/
*pygubu/src/pygubu/plugins/tk/tkstdwidgets.py*
which the program run in virtual environment only

Would you point out which step is wrong


Robinson Law


Alejandro Autalán ***@***.***> 於 2023年1月13日週五 下午12:33寫道：

**Reply by alejandroautalan** [reply]
Hi. You are on Windows OS (the steps will be different). I think you need an xserver for windows.
This [video](https://www.youtube.com/watch?v=BDilFZ9C9mw) can help you.

I will try to make a test on windows tomorrow.

Regards
Alejandro A.

**Comment by RobinsonLaw** [comment, marked as answer]
Hi

1. issue in wrong path of pygubu lib is solved by " RUN pip install -e
./pygubu "
2. *But same error occur that need to ADD DISPLAY environment variable
although windows does not use this feature. I see many people have found
this error in using tk in docker.*

*Robinson*


Alejandro Autalán ***@***.***> 於 2023年1月13日週五 下午3:23寫道：

**Reply by RobinsonLaw** [reply]
Solved By Add

ENV DISPLAY=host.docker.internal:0.0
